### PR TITLE
Add sdk interface specification

### PR DESCRIPTION
**Pull request by iteyelmp**

### CODE REVIEW

File: src/spec.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+This document aims to standardize the interaction between applications and the EthStorage network to achieve reliable and efficient data management functionality.
+
+This document includes two main classes: `EthStorage` and `FlatDirectory`. 
+The `EthStorage` class provides asynchronous read and write operations for key-value pairs of a specified size. 
+The `FlatDirectory` class is a higher-level data management tool that provides methods for uploading and downloading data of arbitrary size.
+
+
+<p id="Terminology"></p>
+
+## 2. Terminology
```
Comment (qzhodl) [inline src/spec.md:38]: Suggest to remove this part

File: src/spec.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+### Constructor
+
+**Description**: Create an instance of `EthStorage`.
+
+**Parameters**
+- `config` (object): Configuration object containing necessary settings.
+   - `rpc` (string): RPC for any evm network.
+   - `privateKey` (string): Wallet private key.
+   - `address` (string): EthStorage contract address.
```
Comment (qzhodl) [inline src/spec.md:80]: Is it optional because we know the default contract address for a specified RPC?

File: src/spec.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+### Methods
+
+#### write
+
+**Description**: Asynchronously writes data to the EthStorage network.
+
+**Parameters**
+- `key` (string): The key for the data to be written.
+- `data` (Buffer): The data to be written, its size cannot exceed the maximum value of the content that can be transferred by a blob.
```
Comment (qzhodl) [inline src/spec.md:101]: We may add `putBlobs` to support uploading multiple BLOBs in a transaction

Comment (qzhodl) [inline src/spec.md:101]: People may want to know the gas and storage costs before writing the data

File: src/spec.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+**Parameters**
+- `ethStorageAddress` (string): EthStorage contract address, used to store data in EthStorage.
+
+**Returns**
+- `address` (string): FlatDirectory address.
+
+**Example**
+```javascript
+const ethStorageAddress = "0x804C520d3c084C805E37A35E90057Ac32831F96f";
+const data = await ethStorage.deploy(ethStorageAddress);
```
Comment (qzhodl) [inline src/spec.md:216]: The return variable's name should be an address, not a data

File: src/spec.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+### Methods
+
+#### upload
+
+**Description**: Upload data of arbitrary size.
+
+**Parameters**
+- `key` (string): The key of the data.
+- `data` (Buffer): The data to be uploaded.
```
Comment (qzhodl) [inline src/spec.md:160]: Do we want to support the stream?

File: src/spec.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+};
+
+const flatDirectory = new FlatDirectory(config);
+```
+
+### Methods
+
+#### upload
+
+**Description**: Upload data of arbitrary size.
```
Comment (qzhodl) [inline src/spec.md:156]: We may also wanna know the gas and storage cost before uploading

File: src/spec.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+- `ethStorageRpc` (string): RPC of EthStorage network, because the data is obtained from the EthStorage network.
+
+**Returns**
+- `data` (Buffer): The content.
+
+**Example**
+```javascript
+const ethStorageRpc = "https://xxx.rpc";
+
+const data = await ethStorage.download("example.txt", ethStorageRpc);
```
Comment (qzhodl) [inline src/spec.md:199]: Why doesn’t the download function have a callback function to track the progress?

File: src/spec.md [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+---
+
+<p id="Introduction"></p>
+
+## 1. Introduction
+
+This document aims to standardize the interaction between applications and the EthStorage network to achieve reliable and efficient data management functionality.
+
+This document includes two main classes: `EthStorage` and `FlatDirectory`.
```
Comment (syntrust) [inline src/spec.md:31]: The sdk instead of document includes ...

File: src/spec.md [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+**Description**: Upload data of arbitrary size.
+
+**Parameters**
+- `key` (string): The key of the data.
+- `data` (Buffer): The data to be uploaded.
+- `callbacks` (object): An object containing callback functions:
+  - `onProgress` (function): Callback function that receives `(progress)`, where `progress` is an object containing `count` and `totalCount`.
+  - `onFail` (function): Callback function that receives `(error)`.
+  - `onSuccess` (function): Callback function that receives `()`.
```
Comment (syntrust) [inline src/spec.md:164]: receives `()` Is confusing

File: src/spec.md [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+const data = await ethStorage.deploy(ethStorageAddress);
+```
+
+
+#### setDefault
+
+**Description**: Set the default file for FlatDirectory.
+
+**Parameters**
+- `defaultFile` (string): The filename of the default file.
```
Comment (syntrust) [inline src/spec.md:225]: Better describe what is the purpose of setting a default file to new users?

File: src/spec.md (hunk 5 above)
Comment (iteyelmp) [reply src/spec.md:225]: The browser does not support stream

Comment (qzhodl) [reply src/spec.md:225]: https://stackoverflow.com/a/65087341/2774879

File: src/spec.md [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+<p id="Class_Overview"></p>
+
+## 2. Class Overview
+
+### EthStorage Class
+
+| Method Name  | Description                                      |
+|--------------|--------------------------------------------------|
+| constructor  | Create an instance of EthStorage                 |
+| estimateCost | Estimate the cost of uploading data              |
```
Comment (qzhodl) [inline src/spec.md:49]: Is it including gas cost or not?

File: src/spec.md [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+const data = await ethStorage.read("example.txt");
+```
+
+
+#### putBlobs
+**Description**: Batch upload blob data to the EthStorage network.
+
+**Parameters**
+- `number` (number): Number of blobs.
+- `data` (Buffer): Blob content to be written, its size cannot exceed the maximum value of the content that can be transferred by a blob.
```
Comment (qzhodl) [inline src/spec.md:155]: Since there are multiple blobs, why the size cannot exceed a blob?

File: src/spec.md [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+### Constructor
+
+**Description**: Create an instance of `FlatDirectory`.
+
+**Parameters**
+- `config` (object): Configuration object containing necessary settings.
+    - `rpc` (string): RPC for any evm network.
+    - `ethStorageRpc` (string): The EthStorage network rpc corresponding to this evm network, the data is obtained from the EthStorage network.
+    - `privateKey` (string): Wallet private key.
+    - `address` (string, optional): FlatDirectory contract address. If it does not exist, the `deploy` method can be called to create one.
```
Comment (qzhodl) [inline src/spec.md:181]: If it does not exist, the `deploy` method need to be called before calling other function?

Comment (qzhodl) [reply src/spec.md:181]: If the answer is yes, I would suggest that moving the explanation of the deploy function to the first one.

**Review by qzhodl — APPROVED** [review]

File: src/spec.md [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+- `number` (number): Number of blobs.
+- `data` (Buffer): Blob content to be written.
+
+**Returns**
+- `status` (Promise<boolean>): A Promise that resolves to the execution result. `true|false`
+
+**Example**
+```javascript
+const blobData = Buffer.from("some data");
+const status = await ethStorage.putBlobs(number, blobData);
```
Comment (syntrust) [inline src/spec.md:166]: Is it possible for the SDK to calculate the number for the user based on the blob data?

File: src/spec.md [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+
+**Parameters**
+- `key` (string): The key for the data to be read.
+
+**Returns**
+- `data` (Promise<Buffer>): A Promise that resolves to the downloaded data.
+
+**Example**
+```javascript
+const data = await ethStorage.downloadSync("example.txt");
```
Comment (syntrust) [inline src/spec.md:310]: It appears that the examples are misplaced for download and downloadSync.

File: src/spec.md (hunk 14 above)
Comment (iteyelmp) [reply src/spec.md:166]: It's more difficult. Why want to do this?

Comment (syntrust) [reply src/spec.md:166]: Maybe it is also difficult for the user?

Comment (iteyelmp) [reply src/spec.md:166]: The SDK only provides support for the contract function `function putBlobs(uint256 num) public payable`.  
The function to calculate `num` should not be placed in the SDK, but where the SDK is called.

**Review by syntrust — APPROVED** [review]

**Pull request merged**